Add Zulip specific fabric instructions

hackerkid opened the pull request:
This PR has the minimal amount of changes to make the ` digitalocean/marketplace-partners` scripts generate Zulip images. A lot of things can be cleaned up like in https://github.com/zulip/zulip/pull/11622 but has been ignored to make the review process easy.

To test this. Create a fresh droplet in DO.  Then execute the following from the local computer.
```bash
# works only on python2
pip install 'fabric<2.0'
cd marketplace_docs/templates/Fabric/
fab build_test -H <droplet_ip>
```

This will setup the Zulip one click app but will not perform a cleanup of the build system.  After the setup is completed SSH into the droplet and try out the one click installer.

If everything looks good create **a new droplet** and run
```
fab build_image -H <droplet_ip>
```
This task setup one click app as well as clean up the build system to prepare your droplet for snapshot. After that image can be created from DO dashboard after shutting down the droplet.

**Note** Currently this works only with python2. I have made modifications to make this work in Python3 in https://github.com/zulip/zulip/pull/11622. I will create a seperate PR with modifications once we get this merged and deploy the image in marketplace.

hackerkid committed:
gitignore: Add python, pip and virtualenv files to gitignore.

hackerkid committed:
Create Zulip specfic fabric instructions for generating image.

timabbott commented:
Cool, just to be clear, the Python2+3 improvements are things we'd submit upstream?

timabbott commented on marketplace_docs/templates/Fabric/files/etc/update-motd.d/99-one-click:
```diff
@@ -2,14 +2,12 @@
-This is the text that will be shown in the MOTD users see when they log in via
-SSH.  This copy should include links to documentation or product informaiton
-along with any information that would helpful to users.
+If you have any queries visit https://chat.zulip.org to get help from the Zulip community.
```
We should probably link to the ReadTheDocs troubleshooting guide instead, since that has some basic guidelines like that they should go to `#production help`.

timabbott commented on marketplace_docs/templates/Fabric/files/etc/update-motd.d/99-one-click:
```diff
@@ -2,14 +2,12 @@
-SSH.  This copy should include links to documentation or product informaiton
-along with any information that would helpful to users.
+If you have any queries visit https://chat.zulip.org to get help from the Zulip community.
+For enterprise plans and support mail at support@zulipchat.com.
```
I think linking to https://zulipchat.com/plans is probably better than the email address.

@rishig any other thoughts on this text, which is what will show up every time folks logic to systems created using this 1-click installer?  (So pretty highly visible)

hackerkid replied:
@timabbott Yup. They have recommended it like that. https://github.com/zulip/marketplace-partners/blob/master/marketplace_docs/build-an-image-fabric.md#running-commands-on-first-login

timabbott commented on marketplace_docs/templates/Fabric/fabfile.py:
```diff
@@ -102,6 +102,12 @@ def run_scripts():
+    put("interactive_script.sh", "/opt/zulip/interactive_script.sh")
+    run("chmod +x /opt/zulip/interactive_script.sh")
+    run("cp /root/.bashrc /etc/skel/.zulip_bashrc")
+    run("echo '/opt/zulip/interactive_script.sh' >> /root/.bashrc")
```
I assume that this approach of putting something in `.bashrc` is the recommended approach for an interactive script?  Seems janky, but OK if that's what they recommend.

hackerkid replied:
Yup. They have recommended it like that. https://github.com/zulip/marketplace-partners/blob/master/marketplace_docs/build-an-image-fabric.md#running-commands-on-first-login

timabbott commented:
OK, posted a batch of comments.  Overall this looks pretty good.  Just to make sure I understand, one would probably just run `build_image` when we're releasing a new version, right?

I think after these modifications are done (especially the MOTD text), I'm happy to merge this and submit it to Digital Ocean's store.

timabbott commented:
Also, I think it's worth extracting from https://github.com/zulip/zulip/pull/11622 the documentation for managing/updating this, and put it in an article in the "subsystems" section of ReadTheDocs, with a link in https://zulip.readthedocs.io/en/latest/subsystems/release-checklist.html

timabbott commented:
Finally, this is exciting!  Thanks for all of your work on this, and sorry my review has been slow.

hackerkid commented:
> OK, posted a batch of comments. Overall this looks pretty good. Just to make sure I understand, one would probably just run build_image when we're releasing a new version, right?

@timabbott Yup. Thats right. We will run `build_image` on a new droplet and then shutdown and create image for it.

hackerkid commented:
> Cool, just to be clear, the Python2+3 improvements are things we'd submit upstream?

Yeah. We can do that. Since Python2 EOL is approaching they might also want to merge that. The only issue is that the official fabric package that runs on Python3 has a vastly diffrent API. So we need to either do a complete rewrite for the new API or use fabric3 a third party package that runs on Python3 and has the same API as Fabric package that runs in python2.7. The later way is how I made it work using Python3. The DO docs also recommend using Fabric3. I can create a PR to that up upstream later.

timabbott commented:
OK, great, do a pass of updating the MOTD and then @rishig and/or I can do a second edit pass on it and then we can merge.

hackerkid commented:
Updated.

rishig commented on marketplace_docs/templates/Fabric/files/etc/update-motd.d/99-one-click:
```diff
@@ -2,14 +2,12 @@
-SSH.  This copy should include links to documentation or product informaiton
-along with any information that would helpful to users.
+If you have any queries or issues see https://zulip.readthedocs.io/en/latest/production/install.html#troubleshooting.
+For enterprise plans and support visit https://zulipchat.com/plans/
```
Not sure what the text here should be, but maybe the following?

Troubleshooting guide: https://zulip.readthedocs.io/en/stable/production/troubleshooting.html
Secure, maintain, and upgrade: https://zulip.readthedocs.io/en/stable/production/maintain-secure-upgrade.html
Enterprise plans and support: https://zulipchat.com/plans/

timabbott replied:
This looks great except that it's probably going to generate too-long lines since this is a text format.  Maybe just

You may find the following resources helpful:
* https://zulip.readthedocs.io/en/stable/production/troubleshooting.html
* https://zulip.readthedocs.io/en/stable/production/maintain-secure-upgrade.html
* https://zulipchat.com/plans/

rishig replied:
How about this?
```
For troubleshooting, maintainance, and enterprise support plans, see
* https://zulip.readthedocs.io/en/stable/production/troubleshooting.html
* https://zulip.readthedocs.io/en/stable/production/maintain-secure-upgrade.html
* https://zulipchat.com/plans/
```

rishig commented on marketplace_docs/templates/Fabric/interactive_script.sh:
```diff
@@ -0,0 +1,43 @@
+   bash $(readlink -f "${0}")
+
+Please press enter when you are ready to configure Zulip.
+
```
These text snippets are tough -- I haven't installed enough software recently to have a good sense of what's normal. But one pass:

```
Welcome! Zulip will ask for the following two pieces of information during setup:
- An email address for support and error emails, like ops@example.com
- A hostname for the Zulip server, like chat.example.com

To skip Zulip setup for now, press Ctrl+C.  You will be prompted again on your next login.

You can also re-run this setup script at any time with the command
   bash $(readlink -f "${0}")

Press Enter to continue setup.
```

timabbott replied:
This looks totally reasonable.  I would put this somewhere in there (maybe just before the "To skim ..." line.

"For more information on these options, check out:
     https://zulip.readthedocs.io/en/latest/production/install.html#installer-options
"

rishig replied:
they won't get the 2 cert-related options though, right? Just email and hostname?

timabbott replied:
This installer just uses certbot by default.  

Which maybe means these docs should tell them to first register the hostname in DNS or otherwise mention that?

rishig replied:
yeah, that seems reasonable.

Also I'd be worried about linking to a doc with 4 installer options, and then the user feeling like something went wrong when they only see 2 of the 4 during setup?

Maybe we can add a line to that rtd article to clarify (e.g. "Automatic installers only give the first two options.")

rishig commented on marketplace_docs/templates/Fabric/interactive_script.sh:
```diff
@@ -0,0 +1,43 @@
+read -r _
+
+while  [ -z "${email}" ] && [ -z "${hostname}" ]; do
+read -r -p "Enter the email address for recieving support and error emails (ex. user@example.com) " email
```
Email address for support and error emails (e.g. ops@example.com):

timabbott replied:
~~Maybe "Administrator email address to receive support and error emails"?~~ Never mind, was reading from the bottom.

rishig commented on marketplace_docs/templates/Fabric/interactive_script.sh:
```diff
@@ -0,0 +1,43 @@
+
+while  [ -z "${email}" ] && [ -z "${hostname}" ]; do
+read -r -p "Enter the email address for recieving support and error emails (ex. user@example.com) " email
+read -r -p "Enter the domain or subdomain pointed to this Zulip instance (ex. chat.example.com): " hostname
```
Hostname (e.g. chat.example.com):

rishig commented on marketplace_docs/templates/Fabric/interactive_script.sh:
```diff
@@ -0,0 +1,43 @@
+
+cat <<EOM
+
+Zulip instance is now being configured... This might take a few minutes.
```
Configuring Zulip. This might take a few minutes.

rishig reviewed (commented):
Looks great, pretty excited about this! I made some suggested text edits (though maybe get sign-off from Tim first, on the edits).

hackerkid commented:
@timabbott Can you sign off these suggestions?

timabbott commented:
Generally they look great, posted a few more suggests.  It'd be great if you can make these changes and then post a code block or screenshot showing what it actually looks like for a user in the revised version.

hackerkid commented:
Updated the instructions.

## Screen when user login for first time after installing the app.
![Screenshot from 2019-04-15 00-13-01](https://user-images.githubusercontent.com/7190633/56097584-076bca00-5f14-11e9-9085-dfd8ddc1772b.png)

## After pressing enter and entering details
![Screenshot from 2019-04-15 00-13-28](https://user-images.githubusercontent.com/7190633/56097583-076bca00-5f14-11e9-9112-c91485150a17.png)

## When user logins to SSH after the setup is completed.
![Screenshot from 2019-04-15 00-15-43](https://user-images.githubusercontent.com/7190633/56097582-076bca00-5f14-11e9-9c3f-f74f3432bf2b.png)

rishig commented:
cool, that's helpful. A few thoughts

* If they need to register the hostname in DNS with the IP address of the newly created droplet, it won't really feel like a 1-click installer, right? Should we provide a default hostname that's whatever Digital Ocean creates by default? 

* Is there a way to get rid of the Ubuntu stuff above our MOTD?

Also, some edits, now that I see this live:

```
Welcome to Zulip setup! You will need:
- An email address for support and error emails, like ops@example.com. You can change 
  this email later.
- A hostname for the Zulip server, like chat.example.com. The hostname must be registered 
  in DNS [...]

Press Ctrl+C to exit setup. You will be prompted again on your next login.

Press Enter to continue.
```

This is under the assumption that "I want to cancel now and run it again later" situation is uncommon. Also if both the email and hostname can be changed later, I feel like we should just say that and point them to the documentation at the end.

timabbott commented:
> If they need to register the hostname in DNS with the IP address of the newly created droplet, it won't really feel like a 1-click installer, right? Should we provide a default hostname that's whatever Digital Ocean creates by default?

I don't think Digital Ocean exposes this to the machine -- this was one of the first questions I asked DO.

> Is there a way to get rid of the Ubuntu stuff above our MOTD?

Yes, we'd just need to RM the file for it in `/etc/`.   I also kinda think we want a line at the top like "Welcome to your Zulip server" or something, it feels a bit abrupt.

A few other thoughts on seeing this:
* I actually preferred the "Press Ctrl+C to skip setup for now" language.  I thought that was clearer.
* I agree we probably don't need to offer the command to rerun the setup script.
* s/enterprise support/support/ I think is cleaner and more future-proof.
* We should figure out what's up with the `manpath` error and fix it; I think the issue is we need to install locales on the system or something?  Unless that's a problem only for a custom locale you're using.

hackerkid commented:
Updated.

## Initial prompt


![1](https://user-images.githubusercontent.com/7190633/56135636-cb397780-5fae-11e9-8c35-d9a3cfdcba34.png)

##  User login screen after the installation is complete

![Screenshot from 2019-04-15 21-24-06](https://user-images.githubusercontent.com/7190633/56148643-5a06be00-5fc8-11e9-9e68-575d0186c949.png)

## I added some extra code to point to useful information in case the installation fails for some reason


![Screenshot from 2019-04-15 23-03-24](https://user-images.githubusercontent.com/7190633/56153150-1cf3f900-5fd3-11e9-8d40-c0fce95d8734.png)

hackerkid commented on marketplace_docs/templates/Fabric/interactive_script.sh:
```diff
@@ -0,0 +1,47 @@
+array=(./zulip-server-*)
+"${array[0]}"/scripts/setup/install --certbot --email="$email" --hostname="$hostname"
+failed=${PIPESTATUS[0]}
+if [ "$failed" = 1 ]; then
```
(extra code added to show some useful info in case of failure)

@timabbott Please review this code (line 38 - 47) as well when you merge :)

hackerkid commented:
@timabbott 
> We should figure out what's up with the manpath error and fix it; I think the issue is we need to install locales on the system or something? Unless that's a problem only for a custom locale you're using.

We are setting the locale before creating the image
https://github.com/zulip/marketplace-partners/blob/e3c29d6b0408519274066eb61522abcce0cf0e0a/marketplace_docs/templates/Fabric/scripts/01-initial-setup#L12

I am getting that error in all most all the droplets I have created recently. I think its mostly an issue with my locale.

timabbott commented:
Merged as 87e1a2b0c89e7cf9d69390b85be6975d698a6116, after a few small text tweaks and fixing the `$?` issue mentioned above.

I think a follow-up project we're likely to end up doing is to not require them to register DNS first -- e.g. we could instead have a prompt or do a self-signed cert if certbot fails.  But I'm tempted to wait for user feedback from folks using this before we do that, in case I'm wrong.

I think I'm happy for us to request this being posted in their store at this point -- I actually don't know what the steps for that are, let me know if you need me to help with anything or you can just reply to the email thread with what they need.  Thanks @hackerkid!
